src/dshould/ShouldType.d

```diff
@@ -0,0 +1,218 @@
+        terminateChain;
+
+        if (!condition)
+        {
+            throw new FluentException("test failed" ~ left, right, file, line);
+        }
+    }
+
+    public void allowOnlyWordsBefore(string[] allowedWords, string newWord)()
+    {
+        static foreach (word; words)
+        {
+            static assert(allowedWords.canFind(word), `bad grammar: "` ~ word ~ ` ` ~ newWord ~ `"`);
+        }
+    }
+
+    public void terminateChain()
+    {
+        this.refs = CHAIN_TERMINATED; // terminate chain, safe ref checker
+    }
+
+    public void requireWord(string requiredWord, string newWord)()
+    {
+        static assert(
+            words.canFind(requiredWord),
+            `bad grammar: expected "` ~ requiredWord ~ `" before "` ~ newWord ~ `"`
+        );
+    }
+
+    public enum hasWord(string word) = words.canFind(word);
+
+    // work around https://issues.dlang.org/show_bug.cgi?id=18839
+    public auto empty()() { return this.empty_; }
+
+    private enum CHAIN_TERMINATED = int.max;
+
+    private @property ref int refs()
+    {
+        if (this.refs_ is null)
+        {
+            this.refs_ = new int;
+            *this.refs_ = 0;
+        }
+        return *this.refs_;
+    }
+}
+
+// must be here due to https://issues.dlang.org/show_bug.cgi?id=18839
+void empty_(Should)(Should should, string file = __FILE__, size_t line = __LINE__)
+if (isInstanceOf!(ShouldType, Should))
+{
+    import std.range : empty;
+
+    with (should)
+    {
+        allowOnlyWordsBefore!(["be", "not"], "empty");
+        requireWord!("be", "empty");
+
+        terminateChain;
+
+        auto got = should.got();
+
+        static if (hasWord!"not")
+        {
+            check(!got.empty, `: expected nonempty array`, null, file, line);
+        }
+        else
+        {
+            check(got.empty, `: expected empty array`, format(", but got %s", got), file, line);
+        }
+    }
+}
+
+class FluentExceptionImpl(T : Exception) : T
+{
+    private const string leftPart = null; // before reason
+    public const string reason = null;
+    private const string rightPart = null; // after reason
+
+    public this(string leftPart, string rightPart, string file, size_t line) pure @safe
+    {
+        this.leftPart = leftPart;
+        this.rightPart = rightPart;
+
+        super(genMessage, file, line);
+    }
+
+    public this(string leftPart, string reason, string rightPart, string file, size_t line) pure @safe
+    {
+        this.leftPart = leftPart;
+        this.reason = reason;
+        this.rightPart = rightPart;
+
+        super(genMessage, file, line);
+    }
+
+    public this(string msg, string file, size_t line) pure @safe
+    {
+        this.leftPart = msg;
+
+        super(genMessage, file, line);
+    }
+
+    public FluentException because(string reason) pure @safe
+    {
+        return new FluentException(this.leftPart, reason, this.rightPart, this.file, this.line);
+    }
+
+    private string genMessage() pure @safe
```

Comment:
_To name a function, use a description of the return value._:
drop `gen` - no verb phrase required